### ISSUE TITLE
[FEATURE REQUEST]: Support alternating acqf optimization

### ISSUE THREAD

**Issue by TobyBoyne** [experimental-design/bofire#579]
### Motivation

Currently, performing Bayesian optimization over mixed spaces in BoFire requires enumeration over all of the discrete combinations, and optimizing each subproblem. This is expensive, and becomes infeasible when the number of combinations explodes combinatorially. For example, the Cell Network problem in the [Prob Reparam](https://arxiv.org/pdf/2210.10199) paper has $6^{15}$ combinations.

BoFire should support the alternating approach for optimizing the acqf proposed in Wu et al. (2021).

Wu et al. 2021 [Think Global and Act Local: Bayesian Optimisation over High-Dimensional Categorical and Mixed Search Space](https://arxiv.org/pdf/2102.07188)

### Describe the solution you'd like to see implemented in BoFire.

In the `BoTorchOptimizer` class, we should allow the user to select the `optimize_acqf_mixed_alternating` [1] approach to optimize the acquisition function.  https://github.com/experimental-design/bofire/blob/03b1f320c62d0af990e66268d16f23007c34c35f/bofire/strategies/predictives/acqf_optimization.py#L433

[1] https://botorch.readthedocs.io/en/stable/optim.html#botorch.optim.optimize_mixed.optimize_acqf_mixed_alternating

### Describe any alternatives you've considered to the above solution.

More complicated methods, such as Probabilistic Reparameterization, would be great once they have been merged into BoTorch. Simpler methods, such as continuous relaxation, have been shown to not have great performance.

### Pull Request

Yes

**Comment by TobyBoyne** [comment]
Hi @jduerholt,

You are right that the method doesn't support categoricals. I will check what Ax does and get back to you.

Even though this function only works for discrete, is it not still worth adding to BoFire in the case that all your inputs are continuous/discrete/binary? Or do you think we should only add the alternating approach to BoFire once it supports categoricals?

Thanks,
Toby

**Comment by jduerholt** [comment]
Let us have a look how Ax is doing it and then decide. @LukasHebing also added an GA for highly categorical search spaces, did you already had a look at it?

**Comment by LukasHebing** [comment]
I didn't do any studies on that topic. As far as I can say is, that GAs are usually quite robust in all search spaces. This is why we chose a GA for our app, where we don't see which search spaces the users provide. 

However, in comparison to gradient-based optimizations, the GA is really inefficient. So it might be good to take a look at other mixed-integer algorithms.

**Comment by TobyBoyne** [comment]
@jduerholt I'm not very familiar with Ax, so I'm not quite sure if I'm understanding correctly their approach. However, from looking through some closed issues, and looking at their code, they enumerate all of the possible categorical combinations. If that results in too many combinations, they fall back to Sobol sampling.

https://github.com/facebook/Ax/blob/543a58302c4e10a51f0edddcd283ced812615112/ax/modelbridge/dispatch_utils.py#L209-L215

Thank you for pointing me towards the GA approach, I'll have a look!

**Comment by jduerholt** [comment]
Hi @TobyBoyne,

I just saw that you manged to get the categoricals into BoTorch! Congrats! This is really great!!!!!

It would be really nice to make use of it in BoFire ;) But it could be that we need also some changes regarding the `input_preproccesing_specs` and one hot, categorical transforms etc. 

From looking at the code, it assume that categoricals are always integer encoded, or? In BoFire it is always one-hot or categorical encoded. We would then need to do this transforms in the GP, or? See also this discussion: https://github.com/experimental-design/bofire/pull/476#issuecomment-2538693780

Should we schedule a meeting to discuss how to approach it? I have the feeling that this will end in some larger refactoring on the bofire side, which I have in mind already for a longer time ;)

Still missing would be in the botorch also:
- Linear equality constraints
- Nonlinear constraints

But I will add those ;)

Best,

Johannes

**Comment by TobyBoyne** [comment]
Hi @jduerholt,

Thank you :)

So the new code in BoTorch, yes it assumes an ordinal integer encoding for the categoricals. You could also use the `optimize_acqf_mixed_alternating` function with one-hot encoding and the constraints `sum(x) <= 1 & sum(x) >= 1`, which should have the same result.

I did try to get this working with BoFire, but I did have some issues with how InputTransforms are handled.

I have emailed you to discuss a meeting!
Thanks,
Toby
